src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,952 @@
+
+  let join a b =
+    let res = join a b in
+    if M.tracing then M.tracel "join" "join a: %s b: %s -> %s \n" (show a) (show b) (show res) ;
+    res
+  let widen a b =
+    let a_env = a.env in
+    let b_env = b.env in
+    if Environment.equal a_env b_env  then
+      join a b
+    else b
+
+  let remove_rels_with_var x var env imp =
+    let j0 = Environment.dim_of_var env var in
+    if imp then (EArray.reduce_col_with x j0; x) else EArray.reduce_col x j0
+
+  let narrow a b = a
+  let pretty_diff () (x, y) =
+    dprintf "%s: %a not leq %a" (name ()) pretty x pretty y
+
+  let forget_vars t vars = 
+    if is_bot_env t || is_top t then t
+    else
+      let m = Option.get t.d in
+      if List.is_empty vars then t else
+        let rec rem_vars m vars' =
+          begin match vars' with
+            |            [] -> m
+            | x :: xs -> rem_vars (remove_rels_with_var m x t.env true) xs end 
+        in {d = Some (EArray.remove_zero_rows @@ rem_vars (EArray.copy m) vars); env = t.env}
+
+  let forget_vars t vars =
+    let res = forget_vars t vars in
+    if M.tracing then M.tracel "ops" "forget_vars %s -> %s\n" (show t) (show res);
+    res
+
+  let forget_vars t vars = timing_wrap "forget_vars" (forget_vars t) vars
+  (* implemented as described on page 10 in the paper about Fast Interprocedural Linear Two-Variable Equalities in the Section "Abstract Effect of Statements" 
+     This makes a copy of the data structure, it doesn't change it in-place. *)
+  let assign_texpr (t: VarManagement.t) var texp =
+    let assigned_var = Environment.dim_of_var t.env var  (* this is the variable we are assigning to *) in
+    begin match t.d with 
+      | Some d -> 
+        let abstract_exists_var = abstract_exists var t in
+        begin match get_coeff t texp with
+          | None -> (* Statement "assigned_var = ?" (non-linear assignment) *) abstract_exists_var
+          | Some (exp_var_opt, off) -> 
+            begin match exp_var_opt with
+              | None -> (* Statement "assigned_var = off" (constant assignment) *) 
+                assign_const abstract_exists_var assigned_var off
+              | Some exp_var (* Statement "assigned_var = exp_var + off" (linear assignment) *) 
+                -> begin if assigned_var = exp_var then 
+                      (* Statement "assigned_var = assigned_var + off" *)
+                      subtract_const_from_var t assigned_var off
+                    else
+                      (* Statement "assigned_var = exp_var + off" (assigned_var is not the same as exp_var) *) 
+                      let empty_array = EqualitiesArray.make_empty_array (VarManagement.size t) in
+                      let added_equality = empty_array.(assigned_var) <- (Some exp_var, off); empty_array in
+                      meet abstract_exists_var {d = Some added_equality; env = t.env} 
+                  end 
+            end
+        end 
+      | None -> bot_env end 
```

Comment:
Could one reduce the level of nesting here by combining some of the `match` statements?